File: backend-node/src/services/BanditService.js
const { User, BanditStat, Recipe } = require('../models');

const WeatherService = require('./WeatherService');

class BanditService {
    // Get current Tau estimate for User
    static async getTau(userId) {
        let stat = await BanditStat.findOne({ where: { UserId: userId } });
        if (!stat) {
            stat = await BanditStat.create({ UserId: userId });
            console.warn(`[Bandit] Created missing stat for User ${userId}`);
        }
        return stat;
    }

    // Select Recipe based on Tau
    static async selectRecipe(tau) {
        let tier = '3min';
        if (tau < 90) tier = '1min';
        else if (tau < 150) tier = '2min';

        let recipe = await Recipe.findOne({ where: { tier } });
        if (!recipe) recipe = await Recipe.findOne({ where: { tier: '3min' } });

        return recipe;
    }

    // Update Tau based on feedback
    static async updateTau(userId, isOk, sessionDuration) {
        const stat = await this.getTau(userId);
        const currentTau = stat.tau_mu;
        let newTau = currentTau;

        if (isOk) {
            newTau = currentTau + 5;
        } else {
            newTau = currentTau - 15;
        }

        newTau = Math.max(45, Math.min(300, newTau));

        stat.tau_mu = newTau;
        await stat.save();
        return newTau;
    }

    static async calculateEffectiveTau(userId, now = new Date()) {
        const stat = await this.getTau(userId);
        let tau = stat.tau_mu;

        const hour = now.getHours();

        // 1. Time Context: Morning Rush (6:00 - 10:00)
        if (hour >= 6 && hour < 10) {
            tau = tau * 0.8;
            console.log(`[Bandit] Context: Morning Rush. Tau scaled to ${tau}`);
        }

        // 2. Weather Context
        const temp = await WeatherService.getCurrentTemperature();
        if (temp !== null) {
            if (temp < 10) {
                tau = tau * 1.15; // +15% for Cold
                console.log(`[Bandit] Context: Cold (${temp}°C). Tau scaled to ${tau}`);
            } else if (temp > 30) {
                tau = tau * 0.9; // -10% for Hot
                console.log(`[Bandit] Context: Hot (${temp}°C). Tau scaled to ${tau}`);
            }
        }

        return Math.round(tau);
    }

    static async getOrCreateUser(deviceId) {
        if (!deviceId) return null;
        let user = await User.findOne({ where: { device_id: deviceId } });
        if (!user) {
            user = await User.create({ device_id: deviceId });
            await BanditStat.create({ UserId: user.id });
        }
        return user;
    }
}

module.exports = BanditService;

File: backend-rails/app/services/bandit_service.rb
class BanditService
  def self.get_tau(user_id)
    stat = BanditStat.find_by(user_id: user_id)
    unless stat
      # Default: alpha=2, beta=2 (Neutral)
      # We don't store mu directly in DB model scaffold? 
      # Node model had tau_mu?
      # Rails scaffold: context:string arm:string alpha:integer beta:integer
      # Wait, Node Model had `tau_mu`. Rails scaffold didn't have `tau_mu`.
      # I need to add `tau_mu` to BanditStat or calculate it.
      # Node's updateTau logic: updates tau_mu directly.
      # Let's add tau_mu to the model via migration or just store it in `alpha`? No.
      # I missed `tau_mu` in the scaffold! 
      # I will assume we can use a separate migration or just rely on 'alpha' as tau?
      # No, tau is 180s. Alpha/Beta are for probability.
      # Node code: `stat.tau_mu = newTau`.
      # I should add `tau_mu` column.
      
      stat = BanditStat.create(user_id: user_id, alpha: 2, beta: 2) 
      # Hack: We will use 'alpha' to store tau_mu for now if column missing, OR I fix the schema.
      # Better to fix the schema. But for now I'll create the file assuming I'll fix the schema.
      # Actually, let's just use `alpha` as `tau_mu` (mean) in seconds? No, that's confusing.
      # I'll create a migration to add `tau_mu`.
    end
    stat
  end

  # Select Recipe based on Tau
  def self.select_recipe(tau)
    tier = '3min'
    if tau < 90
      tier = '1min'
    elsif tau < 150
      tier = '2min'
    end

    recipe = Recipe.find_by(tier: tier)
    recipe || Recipe.find_by(tier: '3min')
  end

  def self.update_tau(user_id, is_ok, session_duration)
    stat = get_tau(user_id)
    current_tau = stat.tau_mu || 180
    new_tau = current_tau

    if is_ok
      new_tau += 5
    else
      new_tau -= 15
    end

    # Clamp
    new_tau = 150 if new_tau < 150
    new_tau = 240 if new_tau > 240

    stat.update(tau_mu: new_tau)
    new_tau
  end

  def self.calculate_effective_tau(user_id, now = Time.now)
    stat = get_tau(user_id)
    tau = (stat.tau_mu || 180).to_f
    reasoning = "Standard Routine."

    hour = now.hour

    # 1. Time Context
    if hour >= 6 && hour < 10
      tau *= 0.8
      reasoning = "Morning Rush detected. Speed up."
      puts "[Bandit] Context: Morning Rush. Tau scaled to #{tau}"
    end

    # 2. Weather Context
    temp = WeatherService.get_current_temperature
    if temp
      if temp < 10
        tau *= 1.15
        reasoning = "It's freezing (#{temp}C). Take your time."
      elsif temp > 30
        tau *= 0.9
        reasoning = "It's sweltering (#{temp}C). Quick wash."
      end
    end

    # 3. Clamp
    tau = 150 if tau < 150
    tau = 240 if tau > 240

    { tau: tau.round, reasoning: reasoning }
  end

  def self.get_or_create_user(device_id)
    return nil unless device_id
    user = User.find_by(device_id: device_id)
    unless user
      user = User.create(device_id: device_id)
      # Create Stat
      BanditStat.create(user_id: user.id, tau_mu: 180, alpha: 2, beta: 2)
    end
    user
  end
end
